### PR TITLE
Add feature to show exact error line number when using `istioctl analyze`

### PR DESCRIPTION
**Pull request by hanxiaop**
This PR is to solve issue #22872. The original feature shows only the starting line number of the error resource for the message, for example: 

`Error [IST0101] (VirtualService frontend.default example-vs.json:5) Referenced gateway not found: "httpbin-gateway-bogus"`

If the gateway error place `"httpbin-gateway-bogus"` is actually in line 10, now it will show like below:

`Error [IST0101] (VirtualService frontend.default example-vs.json:10) Referenced gateway not found: "httpbin-gateway-bogus"`

As mentioned in issue #22824, we didn't have enough low-level information to show more accurate number. The implementation is based on the error messages. The entire YAML chunk resource is used to generate a map with all the line numbers mapped to corresponding values. The actual line number of the message is found by using the parameters provided from the message, and the YAML resource map. If in the future more error types are supported, it will take 2-3 minutes to update the keywords that needs to be examined in order for this error type to be supported by this new feature. A design documentation will be shared to the community after discussion and revision.

The next stage of improvement is in progress based on this PR to make the feature more developer-friendly when finding the JSON path key from the map in the analyzers. Once that PR is created, I'll leave that PR number in this one.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: galley/pkg/config/analysis/diag/message.go
Comment (zerobfd) [inline galley/pkg/config/analysis/diag/message.go:82]: I believe the Reference already knows whether it has a line number or not, so you should be able to put the string formatting logic into String() itself and not need a different function.

File: galley/pkg/config/analysis/diag/message.go
Comment (zerobfd) [reply galley/pkg/config/analysis/diag/message.go:180]: Discussed more offline, but if we push this logic down to the analyzers themselves then we can get away with no specific error logic like this and instead just use analyzer specific logic, which can be iterated on to being handled completely automatically.

**Review by zerobfd — CHANGES_REQUESTED** [review]
Overall strategy looks OK. Lots of small details to fix. I think overall it would be much easier to understand if you could do all of the required work while parsing the yaml file and then in the analyzers. That would avoid the need to do a big search in one spot. It should be roughly the same complexity but much simpler.

File: galley/pkg/config/analysis/analyzers/util/find_errorline_utils.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/util/find_errorline_utils.go:27]: I have a few comments on the error functions. I'm going to make one comment, but these apply to all of the functions.

1) All of the other functions take the *resource.Instance as the first argument, can we change these to be the same?
2) I think that these should return (int, ok bool) so that if there's no error line found there's a way to check. In general using a zero value as "not found" isn't good practice, see https://golang.org/doc/effective_go.html#multiple-returns
3) Just return FindErrorLine(...) intead of setting a variable every time.

File: galley/pkg/config/analysis/analyzers/util/find_errorline_utils.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/util/find_errorline_utils.go:150]: As noted above, this should return (int, bool). In that case just
`return FindErrorLine(...)`
in the other functions will return the right values.

File: galley/pkg/config/analysis/analyzers/annotations/annotations.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/annotations/annotations.go:90]: Just export the Line property and set it directly, there's no need to use a setter function.

File: galley/pkg/config/analysis/analyzers/virtualservice/destinationhosts.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/virtualservice/destinationhosts.go:137]: We should find a better way to tell mirror destinations from normal ones.

File: galley/pkg/config/analysis/analyzers/virtualservice/regexes.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/virtualservice/regexes.go:104]: You can combine these cases with a comma, see https://golang.org/doc/effective_go.html#switch

File: galley/pkg/config/analysis/analyzers/virtualservice/util.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/virtualservice/util.go:43]: Unless there's a strong reason to not export the properties, it would be simpler to just export them and get rid of the getter functions. If there's a strong reason, then please remove "Get" from the function names, see https://golang.org/doc/effective_go.html#Getters

File: galley/pkg/config/analysis/analyzers/virtualservice/regexes.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/virtualservice/regexes.go:105]: If we know that the case will be, let's name them specifically and not use default. Default should raise an error.

File: galley/pkg/config/analysis/diag/message.go
Comment (zerobfd) [inline galley/pkg/config/analysis/diag/message.go:143]: Prefer exporting Line property, otherwise get rid of Get in getter name

File: galley/pkg/config/source/kube/inmemory/kubesource.go
Comment (zerobfd) [inline galley/pkg/config/source/kube/inmemory/kubesource.go:367]: Can you flatten this to just be if len(valueNode.Content) == 0 {} else if valueNode.Tag == "!!seq" {} else if valueNode.Tag == "!!map" {}?
If so, please use a switch statement: see https://golang.org/doc/effective_go.html#switch

File: galley/pkg/config/util/kubeyaml/kubeyaml.go
Comment (zerobfd) [inline galley/pkg/config/util/kubeyaml/kubeyaml.go:162]: Why do this? Can we just change the line parameter to be resStr instead?

File: galley/pkg/config/util/kubeyaml/kubeyaml.go
Comment (zerobfd) [inline galley/pkg/config/util/kubeyaml/kubeyaml.go:188]: Do this in one Sprintf

**Review by zerobfd — CHANGES_REQUESTED** [review]

File: galley/pkg/config/source/kube/inmemory/kubesource.go
Comment (monsteredc) [reply galley/pkg/config/source/kube/inmemory/kubesource.go:367]: Thanks!

File: istioctl/cmd/analyze.go
Comment (zerobfd) [inline istioctl/cmd/analyze.go:454]: Should be unnecessary, see comments in message.go

**Review by zerobfd — CHANGES_REQUESTED** [review]

File: galley/pkg/config/analysis/diag/message.go
Comment (monsteredc) [reply galley/pkg/config/analysis/diag/message.go:64]: I think this doesn't work because Origin structure is in the resource.Instance. Different messages may share the same resource.Instance, and this is why the starting line number always works because different messages will show the same line number. However, if we change the line number in the Reference structure to be the exact line, whenever the new message shares the same resource with the former message, the line number in the Reference structure will be replaced to the newer message's error line number.

File: galley/pkg/config/source/kube/inmemory/kubesource.go
Comment (monsteredc) [reply galley/pkg/config/source/kube/inmemory/kubesource.go:330]: I think it doesn't. Only after parsing the YAML chunk correctly and before converting all of the objects to the resource will require this function to build the map. Or if someone else in other places needs to use this function to build a map, then it can be exported.

File: galley/pkg/config/source/kube/inmemory/kubesource.go
Comment (jasonwzm) [reply galley/pkg/config/source/kube/inmemory/kubesource.go:351]: +1

File: go.mod
Comment (jasonwzm) [reply go.mod:162]: We are still using v2 in pilot code.

File: galley/pkg/config/analysis/diag/message.go
Comment (zerobfd) [reply galley/pkg/config/analysis/diag/message.go:64]: Good point. Here's what we can do instead: the Reference interface is only used in analysis, so let's extend it. Right now it only has a String() method. The high level problem we're running into here is that the error message and the line number aren't encapsulated where we need them--the only way we can get the line number out of the Position struct is via String, because that's all the interface has. I propose:

```golang
type Reference interface {
  String() string
  References() map[string]Reference
}
```
Now when we're iterating through the yaml and recursing through all of the Nodes, we can create a Position struct for each level and create a matching structure with Positions. i.e. this:
```yaml
parent:
  levelOne:
    - levelTwo: a
    - levelTwo: b
```
would become something like
```
Position {filename, 0, map[string]Position {
  'levelOne': Position{filename, 1, map[string]Position {
    '0': Position{filename, 2, map[string]Position{}},
    '1': Position{filename, 3, map[string]Position{}},
  },
}
```
Then in the individual ErrorLine functions, instead of building the entire string like
`"parent.levelOne[%d]"`, we can first look for "parent", then recursively look for "levelOne", and so on until we find the key we're looking for (or reach an error). I admit that the handling of arrays as well as maps is a bit inelegant, but we can start from here and make it look nicer later.

This should require minimal refactoring but also allow us to keep the same printing logic (since we're still printing a Reference at the end), and since the line number is part of the struct we don't need to search a string to replace it.

Comment (monsteredc) [reply galley/pkg/config/analysis/diag/message.go:64]: Looks like it's better than writing those utility functions. Thanks a lot for providing this idea! However, can you clarify how to use this nested Reference? From my understanding, the analyzers' developers are going to provide the path variables for the message, and during the message rendering process, the path variables are going to be used to search for the correct Reference.

For example, if a pathKey is originally like "parent.childOne[0].childTwo", if we want to obtain the Reference of childTwo, are we going to pass []interface {"parent", "childOne", 0, "childTwo"} in the analyzer as the variable for the error message, and later use a search function with these variables to find the right Reference to use the String() method?
Is this the correct implementation or these are other ways to finally find the exact Reference?

Comment (zerobfd) [reply galley/pkg/config/analysis/diag/message.go:64]: Yep, that's the basic idea. We could pass the pathKey just like that and split it later--I think you already wrote the logic for parsing the path strings, and I think it would be easier for developers to put a string in instead of a list of strings. i.e. maybe something like

```go
const mirrorPath = "{.spec.http[%d].mirror.host}"
...
func (a *DestinationHostAnalyzer) analyzeVirtualService...
...
if line, ok := util.ErrorLine(r, fmt.Sprintf(mirrorPath, d.ServiceIndex)); ok {
				m.Line = line
			}
```


I don't think we should use interface{} though--just make everything a string. Technically someone could write something like
```yaml
key[3]: value
```
which might not work, but let's not worry about that for the first phase. If it ever becomes an issue we can just use escape codes or something.

File: galley/pkg/config/analysis/analyzers/virtualservice/destinationrules.go
Comment (monsteredc) [inline galley/pkg/config/analysis/analyzers/virtualservice/destinationrules.go:49]: will revise this comment

**Review by jasonwzm — COMMENTED** [review]
We should also add some integration tests in tests/integration/galley/analyze_test.go

**Review by zerobfd — CHANGES_REQUESTED** [review]
I'm pretty much OK with the structure here. Fix the issues, add tests where they're still missing, and it's good to go as long as it satisfies a maintainer.

File: galley/pkg/config/analysis/analyzers/util/find_errorline_utils.go
Comment (monsteredc) [inline galley/pkg/config/analysis/analyzers/util/find_errorline_utils.go:59]: will test all the consts

File: galley/pkg/config/analysis/analyzers/authz/authorizationpolicies.go
Comment (monsteredc) [reply galley/pkg/config/analysis/analyzers/authz/authorizationpolicies.go:171]: The new loop is used to provide the index number of the notNamespaces array for the path key. It can also be done within one loop, however there should be a check condition to the index, like:
`if index >= len(from.Source.Namespaces):  nIndex = index - len(from.Source.Namespaces)`
because namespaces and notNamespaces are two different arrays in the YAML file.

File: galley/pkg/config/analysis/analyzers/virtualservice/util.go
Comment (monsteredc) [reply galley/pkg/config/analysis/analyzers/virtualservice/util.go:21]: Sounds great. Thanks!

File: galley/pkg/config/analysis/diag/message.go
Comment (monsteredc) [reply galley/pkg/config/analysis/diag/message.go:65]: This idea is currently not applied because it only helps the analyzer to display file names and line numbers more easily. However, implementing this idea will make the code structure of the reference part complicated. At the same time, in order to use this structure, we need to parse the complete path into single fields, and then refer to the keys in the Reference map to find the line number. The current method utilizes the string template: the developer only needs to provide the necessary parameters, and when the complete path is available, you can directly search the corresponding line number in the flat field map, which is easier and simpler. However, the idea mentioned above can be pretty useful if there are some simper implementations or different ways to utilize it. It may be more extensible compared to adding string templates.

File: galley/pkg/config/source/kube/inmemory/kubesource.go
Comment (monsteredc) [reply galley/pkg/config/source/kube/inmemory/kubesource.go:361]: This check condition can't be removed because `BuildFieldPathMap` has a for loop with a step of 2. Contents in the node array are like `[key, value, key, value, key, value]` structure. However, in the sequence node, the for loop has an index step of 1. For example, for `gateways` field in `VirtualService` kind, the content of the gateway node is like `[value, value, value]`, and it has a different logic than the `BuildFieldPathMap` function.

File: galley/pkg/config/analysis/analyzers/util/find_errorline_utils.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/util/find_errorline_utils.go:104]: I think you can combine these, findErrorLine doesn't really seem necessary anymore.

File: galley/pkg/config/analysis/analyzers/util/find_errorline_utils_test.go
Comment (zerobfd) [inline galley/pkg/config/analysis/analyzers/util/find_errorline_utils_test.go:95]: Can you just do ```g.Expect(fieldMap[v]).To(Equal(1))```?

File: releasenotes/notes/25302.yaml
Comment (zerobfd) [inline releasenotes/notes/25302.yaml:14]: Let's talk about this in terms of the feature and not what the code does. e.g.

```
*Improved* `istioctl analyze` to find the exact line number with configuration errors when analyzing yaml files. Before, it would return the first line of the resource with the error. 
```

**Review by zerobfd — APPROVED** [review]
A few small things, otherwise it looks alright to me.

File: releasenotes/notes/25302.yaml
Comment (monsteredc) [reply releasenotes/notes/25302.yaml]: Thanks!

**Comment by jasonwzm** [comment]
blocking this PR temporary as https://github.com/istio/istio/pull/25996 is in the queue for merge. This PR will need rebase once 25996 is merged.

**Review by jasonwzm — APPROVED** [review]

File: tests/integration/pilot/analyze_test.go
Comment (therealmitchconnors) [inline tests/integration/pilot/analyze_test.go:346]:
```suggestion
		RequiresSingleCluster().
		Features("usability.observability.analysis.line-numbers").
```
Also, add this constant to features.yaml...

**Review by therealmitchconnors — CHANGES_REQUESTED** [review]

**Review by therealmitchconnors — APPROVED** [review]

**Review by rshriram — APPROVED** [review]

**Pull request merged**
